proposed/http-middleware/middleware.md

```diff
@@ -1,114 +1,123 @@
-interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).
+interpreted as described in [RFC 2119][rfc2119].
+
+[psr7]: http://www.php-fig.org/psr/psr-7/
+[rfc2119]: http://tools.ietf.org/html/rfc2119
 
 ### References
 
-- [RFC 2119](http://tools.ietf.org/html/rfc2119)
-- [PSR-7](http://www.php-fig.org/psr/psr-7/)
+- [PSR-7][psr7]
+- [RFC 2119][rfc2119]
 
 ## 1. Specification
 
-An HTTP middleware component is an individual component participating, together
-with other middleware components, in the processing of an incoming HTTP Request
-and the creation of a resulting HTTP Response, as defined by PSR-7.
+### 1.1 Middleware
 
-Middleware using this standard MUST implement the following interface:
+An HTTP middleware component is an individual component participating, often
+together with other middleware components, in the processing of an incoming
+request and the creation of a resulting response, as defined by PSR-7.
 
-- `Psr\Http\ServerMiddleware\MiddlewareInterface`
+A middleware component MAY create and return a response without delegating to
+a request handler, if sufficient conditions are met.
 
-Middleware dispatching systems using this standard MUST implement the following
-interface:
+Middleware using this standard MUST implement the following interface:
 
-- `Psr\Http\ServerMiddleware\DelegateInterface`
+- `Psr\Http\ServerStrategy\MiddlewareInterface`
 
-Legacy middleware implementing a double pass approach MUST be wrapped using an
-object that implements the `MiddlewareInterface`.
+### 1.2 Request Handlers
 
-### 1.1 Dispatchers
+An HTTP request handler is an individual component that processes a request and
+produces a response, as defined by PSR-7.
 
-An HTTP middleware dispatcher is an object that holds multiple middleware
-components that can be used to process one or more requests in sequence.
+Request handlers using this standard MUST implement the following interface:
 
-The middleware dispatcher MUST pass the request and a delegate to each
-middleware for further processing. The delegate MUST be able to dispatch
-the next available middleware or if no more middleware is available, create a
-default response.
+- `Psr\Http\ServerStrategy\HandlerInterface`
 
-### 1.2 Generating Responses
+### 1.3 Generating Responses
 
-It is RECOMMENDED that any middleware that needs to generate a response will
-use HTTP Factories as defined in [PSR-17](https://github.com/php-fig/fig-standards/tree/master/proposed/http-factory),
-in order to prevent dependence on a specific HTTP message implementation.
+It is RECOMMENDED that any middleware or request handler that generates a response
+will use HTTP factories as defined in [PSR-17][psr17] in order to prevent dependence
+on a specific HTTP message implementation.
+
+[psr17]: https://github.com/php-fig/fig-standards/tree/master/proposed/http-factory
 
 ## 2. Interfaces
 
-### 2.1 Psr\Http\ServerMiddleware\MiddlewareInterface
+### 2.1 Psr\Http\Server\MiddlewareInterface
```

Comment:
Interfaces sub-sections titles and defined namespaces looks not congruent.

Here in 2.1
Psr\Http\\**Server**\MiddlewareInterface
Psr\Http\\**ServerStrategy**\MiddlewareInterface

Idem for HandlerInterface in 2.2